pageserver/src/tenant.rs

```diff
@@ -1394,162 +1438,196 @@ impl Tenant {
+                            // warn! level is technically not appropriate for the
+                            // first case because we should expect retries to happen.
+                            // But the error is so rare, it seems better to get attention if it happens.
+                            let tenant_state = tenant.current_state();
+                            warn!(
+                                timeline_dir=?local_timeline_directory,
+                                ?tenant_state,
+                                "timeline directory not found, proceeding anyway"
+                            );
+                        }
+                        Err(e) => {
+                            warn!(
+                                "failed to remove local timeline directory {}: {e}",
+                                local_timeline_directory.display()
+                            );
+                            return Err(
+                                InnerDeleteTimelineError::FailedToRemoveLocalTimelineDirectory,
+                            );
+                        }
+                    }
+
+                    info!("finished deleting layer files, releasing layer_removal_cs.lock()");
                 }
-                res => res.with_context(|| {
-                    format!(
-                        "Failed to remove local timeline directory '{}'",
-                        local_timeline_directory.display()
-                    )
-                })?,
-            }
 
-            info!("finished deleting layer files, releasing layer_removal_cs.lock()");
-            drop(layer_removal_guard);
-        }
+                fail::fail_point!("timeline-delete-after-rm", |_| {
+                    Err(InnerDeleteTimelineError::Failpoint(
+                        "timeline-delete-after-rm",
+                    ))
+                });
+
+                // Remove the timeline from the map or poison it if we've grown children.
+                let removed_timeline =
+                    std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                        let mut timelines = tenant.timelines.lock().unwrap();
+                        let children_exist = timelines.iter().any(|(_, entry)| {
+                            entry.get_ancestor_timeline_id() == Some(timeline_id)
+                        });
+                        // XXX this can happen because `branch_timeline` doesn't check `TimelineState::Stopping`.
+                        // We already deleted the layer files, so it's probably best to panic.
+                        // (Ideally, above remove_dir_all is atomic so we don't see this timeline after a restart)
+                        if children_exist {
+                            panic!("Timeline grew children while we removed layer files");
+                        }
+                        timelines.remove(&timeline_id)
+                    }));
```

Comment:
1. Noticed the `Ideally, ...` is no longer relevant, so, remove it. It's not atomic but transactional thanks to `deleted_at` tombstone.
2. Move the remaining part of the comment upward, to explain what the effect of the poisoning is.

```suggestion
                // Remove the timeline from the map or poison it if we've grown children.
                // The poisonsed map will make the child's GetPage accesses fail.
                //
                // Growing children can happen because `branch_timeline` doesn't check `TimelineState::Stopping`. 
                let removed_timeline =
                    std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                        let mut timelines = tenant.timelines.lock().unwrap();
                        let children_exist = timelines.iter().any(|(_, entry)| {
                            entry.get_ancestor_timeline_id() == Some(timeline_id)
                        });
                        if children_exist {
                            panic!("Timeline grew children while we removed layer files");
                        }
                        timelines.remove(&timeline_id)
                    }));
```

Reply:
e5fdfae6